src/Raven.Server/Documents/TimeSeries/TimeSeriesPolicyRunner.cs

```diff
@@ -0,0 +1,683 @@
+        {
+            var tss = context.DocumentDatabase.DocumentsStorage.TimeSeriesStorage;
+            if (policy.RetentionTime == null || policy.RetentionTime == TimeSpan.MaxValue)
+                return;
+
+            var deleteFrom = now.Add(-policy.RetentionTime.Value);
+
+            while (true)
+            {
+                Cts.Token.ThrowIfCancellationRequested();
+
+                context.Reset();
+                context.Renew();
+
+                using (context.OpenReadTransaction())
+                {
+                    var list = tss.Stats.GetTimeSeriesByPolicyFromStartDate(context, collectionName, policy.Name, deleteFrom, TimeSeriesRetentionCommand.BatchSize).ToList();
+                    if (list.Count == 0)
+                        return;
+
+                    request.To = deleteFrom;
+                    var cmd = new TimeSeriesRetentionCommand(list, request);
+                    await _database.TxMerger.Enqueue(cmd);
+                }
+            }
+        }
+
+        private void PrepareRollUps(DocumentsOperationContext context, DateTime currentTime, long take, List<RollUpState> states, out Stopwatch duration)
+        {
+            duration = Stopwatch.StartNew();
+
+            var table = context.Transaction.InnerTransaction.OpenTable(RollUpSchema, TimeSeriesRollUps);
+            if (table == null)
+                return;
+
+            var currentTicks = currentTime.Ticks;
+
+            foreach (var item in table.SeekForwardFrom(RollUpSchema.Indexes[NextRollUpIndex], Slices.BeforeAllKeys, 0))
+            {
+                if (take <= 0)
+                    return;
+
+                var rollUpTime = DocumentsStorage.TableValueToEtag((int)RollUpColumn.NextRollUp, ref item.Result.Reader);
+                if (rollUpTime > currentTicks)
+                    return;
+
+                DocumentsStorage.TableValueToSlice(context, (int)RollUpColumn.Key, ref item.Result.Reader,out var key);
+                SplitKey(key, out var docId, out var name);
+
+                var state = new RollUpState
+                {
+                    Key = key,
+                    DocId = docId,
+                    Name = name,
+                    Collection = DocumentsStorage.TableValueToString(context, (int)RollUpColumn.Collection, ref item.Result.Reader),
+                    NextRollUp = new DateTime(rollUpTime),
+                    RollUpPolicy = DocumentsStorage.TableValueToString(context, (int)RollUpColumn.PolicyToApply, ref item.Result.Reader),
+                    Etag = DocumentsStorage.TableValueToLong((int)RollUpColumn.Etag, ref item.Result.Reader),
+                    ChangeVector = DocumentsStorage.TableValueToChangeVector(context, (int)RollUpColumn.ChangeVector, ref item.Result.Reader)
+                };
+
+                states.Add(state);
+                take--;
+            }
+        }
+
+        public static void SplitKey(Slice key, out string docId, out string name)
```

Comment:
Sanity-check. Shouldn't be an issue that we are allocating here instead of 'slicing' Slices?

Reply:
I think that this uses implicit conversions. But should it?

Reply:
We are indeed allocating here. 
But let's optimize it a bit later, since changing this will require to change `string` into `slice` in various places